src/main/java/net/minecraftforge/common/ToolModificationResult.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Minecraft Forge - Forge Development LLC
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.common;
+
+import com.google.common.base.Objects;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Holds a {@link #toolModifiedState() tool-modified state} and whether the modification {@link #passed()} or {@link #failed()}.
+ * <p>
+ * If the tool modification result has a fail state, the tool-modified state is always null.
+ * <p>
+ * If the tool modification result has a pass state, a null tool-modified state signifies a fallback on the default behavior.
+ */
+public final class ToolModificationResult
+{
+    private static final ToolModificationResult FAIL_RESULT = new ToolModificationResult(Type.FAIL, null);
+    private static final ToolModificationResult NULL_PASS_RESULT = new ToolModificationResult(Type.PASS, null);
+    private final Type type;
+    @Nullable
+    private final BlockState toolModifiedState;
+
+    private ToolModificationResult(Type type, @Nullable BlockState toolModifiedState)
+    {
+        this.type = type;
+        this.toolModifiedState = toolModifiedState;
+    }
+
+    public static ToolModificationResult fail()
+    {
+        return FAIL_RESULT;
+    }
+
+    public static ToolModificationResult pass(@Nullable BlockState toolModifiedState)
+    {
+        return toolModifiedState == null ? NULL_PASS_RESULT : new ToolModificationResult(Type.PASS, toolModifiedState);
+    }
+
+    public boolean passed()
+    {
+        return this.type == Type.PASS;
+    }
+
+    public boolean failed()
+    {
+        return this.type == Type.FAIL;
+    }
+
+    public Type type()
+    {
+        return type;
+    }
+
+    /**
+     * If this tool modification result has a {@link #failed() fail state}, this return value is always null.
+     * <p>
+     * If this tool modification result has a {@link #passed() pass state}, a null return value signifies a fallback on the default behavior.
+     *
+     * @return the nullable tool-modified state
+     */
+    @Nullable
+    public BlockState toolModifiedState()
```

Comment:
This is a personal opinion:
Nullable issues are harder to trace and to keep track of.

API wise it would be much better if this method returned an `Optional<BlockState>`

Reply:
I will see what I can do.